### DISCUSSION TITLE
Where is this logged `eval_causal_lm_metrics: ["perplexity"]`?

### DISCUSSION THREAD

**Discussion by nazkhan-8451** [axolotl-ai-cloud/axolotl#1908] [Q&A] [answered]
In my config to finetune llama-3.1, I put `eval_causal_lm_metrics: ["perplexity"].` In `trainer_state.json` here is output at eval step:
```
    "epoch": 0.5002141327623126,
      "eval_loss": 0.8674565553665161,
      "eval_runtime": 562.8213,
      "eval_samples_per_second": 1.747,
      "eval_steps_per_second": 0.437,
      "step": 584
```

Where is `perplexity` logged? I know I can calculate it from cross-entropy loss but wanted to know this.

**Comment by ckgresla** [comment]
I believe that setting this value in your config leads to a perplexity being logged over the training set -- here are a few reference wandb job screenshots from a run with this metric enabled:

<img width="314" alt="training-loss" src="https://github.com/user-attachments/assets/ad63d6a6-f589-43c9-9988-8a66c74260a6">

<img width="410" alt="training-ppl" src="https://github.com/user-attachments/assets/6f1a064f-af29-4ce8-979d-2cd088e449d8">

Additionally, the perplexity here seems oddly high, assuming that the perplexity is calculated as the exponential of the average cross entropy (loss value @ timestep) then we would have:
```
>>> torch.exp(torch.tensor(1.7759))
tensor(5.9056)  # expected perplexity for loss @ timestep
```
but at ~roughly the same timestep, the training perplexity is logged as `~1207` – the implementation of the metric in [perplexity.py](https://github.com/axolotl-ai-cloud/axolotl/blob/main/src/axolotl/utils/callbacks/perplexity.py) seems to be correct, would folks have an intuition as to why this value is higher than the expected perplexity?



I am new to this project and was wondering why the `eval_causal_lm_metrics` get logged only for the training set, the corresponding `eval` section of the wandb run looks like this: 
<img width="1337" alt="Screenshot 2024-09-30 at 12 14 53 PM" src="https://github.com/user-attachments/assets/15a4210b-eeb3-4414-9b4c-25b29446cc8d"> 

Is that the expected behavior, or perhaps is something awry in my setup of the repo? 

Ideally, we could track these casual lm metrics for specified subsets of the data with something like the following:
```
train_causal_lm_metrics: ["perplexity", "accuracy"]
eval_causal_lm_metrics: ["perplexity", "accuracy"]
```


@winglian – would you be open to this becoming a part of the axolotl?

**Reply by NanoCode012** [reply]
Hey @ckgresla, thanks for looking into this. I've been digging into this as well. It turns out that it is indeed from `eval`. It's just that wandb sees it as train split.

If you check this section below, it's within the `CausalLMBenchEvalCallback` during `on_evaluate` callback which `compute` using data from `eval_dataloader`.
https://github.com/axolotl-ai-cloud/axolotl/blob/e1915f5625b2330555c3f61816dd003fb939ae13/src/axolotl/utils/callbacks/__init__.py#L468-L471

We used this to log which had that side effect. https://github.com/axolotl-ai-cloud/axolotl/blob/e1915f5625b2330555c3f61816dd003fb939ae13/src/axolotl/utils/callbacks/__init__.py#L553

This PR should address it https://github.com/axolotl-ai-cloud/axolotl/pull/1952

**Comment by NanoCode012** [comment, marked as answer]
Hey @nazkhan-8451 , the perplexity should show up in there with the changes in #1952 . Don't forget to turn on `do_causal_lm_eval: true` to ensure the eval runs.
